File: src/pipeline/zk.py
from typing import Tuple


def commit(
        s: float, *, poly_coeff: list, g: float
) -> Tuple[float, list[float]]:
    """
    Commits transaction data as a polynomial pool.
    """
    poly = 0.
    user_commits = []
    for idx, item in enumerate(poly_coeff):
        current_poly = item * s ** (len(poly_coeff) - idx - 1)
        poly += current_poly
        user_commits.append(g ** current_poly)
    return g ** poly, user_commits


def witnesses(
        s: float, *, poly_coeff: list, g: float  
) -> list[float]:
    """
    Produces witness parameters for individual data holders.
    """
    witness_array = []
    for idx in range(len(poly_coeff)):
        poly = 0.
        for idx_0, item in enumerate(poly_coeff):
            if idx != idx_0:
                poly += item * s ** (len(poly_coeff) - idx_0 - 1)
        witness_array.append(g ** poly)
    return witness_array


def verify(
        user_commit: float, witness: float, *, agg_commit: float
) -> bool:
    """
    Verifies ownership of data via zk.
    """
    return round(agg_commit / witness, 3) == round(user_commit, 3)

File: frontend/src/zk.js
// Generator of cyclic group G
const g = BigInt(3);
// Secret
const secret = BigInt(7);

const mod = BigInt(1000000000000000000000000000000000) //BigInt(193939);

// Ys is polynomial evaluated at secret point
const commitPoly = (polyCoeff) => {
    let poly = 0n;
    const userCommits = [];
    const elementPoly = [];
    const witnessArray = [];
    const polyLength = polyCoeff.length;

    for (let idx = 0; idx < polyLength; idx++) {
        const currentPoly = (polyCoeff[idx] * ((secret ** BigInt(idx)) % mod)) % mod;
        elementPoly.push(currentPoly);
        poly = (poly + currentPoly) % mod;
        console.log(poly)
        userCommits.push((g ** currentPoly) % mod);
    }

    elementPoly.forEach(element => {
        console.log(g ** (poly - element))

        witnessArray.push((g ** ((poly - element) % mod)) % mod);
    });

    return [(g ** poly) % mod, witnessArray, userCommits];
}

const generateWitnesses = (polyCoeff, s, g) => {
    let witnessArray = [];
    const polyLength = polyCoeff.length;

    for (let idx = 0; idx < polyLength; idx++) {
        let poly = 0n;
        for (let idx_0 = 0; idx_0 < polyLength; idx_0++) {
            if (idx !== idx_0) {
                const currentPoly = (polyCoeff[idx_0] * (s ** BigInt(idx_0)) % mod) % mod;
                poly += currentPoly % mod;
            }
        }
        witnessArray.push((g ** poly) % mod);
    }

    return witnessArray;
}

const verify = (userCommit, witness, aggCommit) => {
    console.log(aggCommit)
    console.log(witness)
    console.log((witness * userCommit) % mod)
    return ((witness * userCommit) % mod) == (aggCommit)
};

export {g, secret, commitPoly, generateWitnesses, verify};